azure/services/groups/groups.go

```diff
@@ -55,65 +60,45 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "groups.Service.Reconcile")
 	defer span.End()
 
-	if _, err := s.client.Get(ctx, s.Scope.ResourceGroup()); err == nil {
-		// resource group already exists, skip creation
-		return nil
-	} else if !azure.ResourceNotFound(err) {
-		return errors.Wrapf(err, "failed to get resource group %s", s.Scope.ResourceGroup())
-	}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-	s.Scope.V(2).Info("creating resource group", "resource group", s.Scope.ResourceGroup())
-	group := resources.Group{
-		Location: to.StringPtr(s.Scope.Location()),
-		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-			ClusterName: s.Scope.ClusterName(),
-			Lifecycle:   infrav1.ResourceLifecycleOwned,
-			Name:        to.StringPtr(s.Scope.ResourceGroup()),
-			Role:        to.StringPtr(infrav1.CommonRole),
-			Additional:  s.Scope.AdditionalTags(),
-		})),
-	}
+	groupSpec := s.Scope.GroupSpec()
 
-	_, err := s.client.CreateOrUpdate(ctx, s.Scope.ResourceGroup(), group)
-	if err != nil {
-		return errors.Wrapf(err, "failed to create resource group %s", s.Scope.ResourceGroup())
-	}
-
-	s.Scope.V(2).Info("successfully created resource group", "resource group", s.Scope.ResourceGroup())
-	return nil
+	err := async.CreateResource(ctx, s.Scope, s.client, groupSpec, serviceName)
+	s.Scope.UpdatePutStatus(infrav1.ResourceGroupReadyCondition, serviceName, err)
+	return err
 }
 
-// Delete deletes the resource group with the provided name.
+// Delete deletes the resource group if it is managed by us.
 func (s *Service) Delete(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "groups.Service.Delete")
 	defer span.End()
 
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
```

Comment:
We should think about handling the cancellation signal here. Right now, we are just cleaning up the resources, but we should also think about what will happen if client.Get for eg. takes longer than "DefaultAzureServiceReconcileTimeout".  It's not going to break anything, but we are not returning even if it exceeds the timeout which was the original intention. wdyt?

Reply:
> but we are not returning even if it exceeds the timeout which was the original intention

can you elaborate on this? 

maybe https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1684/files#diff-5bb14c98c52ecfa3dad6cf4745c9378b7624fdc60f0910343534ddac18408fb1R62 is a better illustration of how this works not just for creates... if the context expires in this function it will return an error which will cause that reconcile loop to fail. For long-running operations, like deletes and creates, we expect that to happen which is why we return a transient "operation not done" error which causes a requeue. If the overall loop times out, eg. a Get takes longer than it should, then maybe it's a good thing to error out and retry later? This is meant to be as an upper bound, ie. this service should never take more than 12 seconds to run the Delete/Reconcile function, unless something went wrong.

Reply:
> if the context expires in this function it will return an error which will cause that reconcile loop to fail

Not sure if that'll happen, though. We'll need to watch for `ctx.Done()` to get the cancellation signal, and return error accordingly. Something similar to what `future.WaitForCompletionRef` does [here](https://github.com/Azure/go-autorest/blob/492f94bc3a562197e229f044dbfabcc0ecb2489a/autorest/azure/async.go#L252). Does that make sense?

Reply:
I see what you're saying now. I guess this isn't really a long-running function where we want to continuously do something and timeout though. The reason for doing a cxt timeout here is that the collection of all the shorter timeouts should never exceed this bigger timeout. When go sets a new timeout deadline, it will select the soonest between the parent context's deadline and the new deadline, so this is meant more as an intermediate timeout (in this case it doesn't really apply though because the group delete only performs one delete operation so it's a 1:1 mapping), same way that the reconcile loop timeout is a global timeout of all the long-running operations we have. Maybe we can discuss this at office hours tomorrow it will be easier to talk it through.

Reply:
Ah I see, so this timeout is just to make sure that the `time(all async operations) < DefaultReconcileTimeout` , and not necessarily `time(service.Reoncile/Delete) < DefaultReconcileTimeout`. I think it makes sense, there shouldn't be any network calls outside of the async interface. ResourceGroups is a special case I guess to check for `IsManaged`, and for other services this will be picked up from a cache.